File: frontend/src/types/llm-usage.ts
/**
 * LLM Usage Analytics Types
 */

export type LLMProvider = 'AZURE_OPENAI' | 'OPENAI' | 'ANTHROPIC'

export type LLMTaskType =
  | 'SUMMARIZE'
  | 'EXTRACT'
  | 'CLASSIFY'
  | 'EMBEDDING'
  | 'VISION'
  | 'CHAT'
  | 'PLAN_MODE'
  | 'DISCOVERY'
  | 'ENTITY_ANALYSIS'
  | 'ATTACHMENT_ANALYSIS'
  | 'RELEVANCE_CHECK'
  | 'CUSTOM'

export type BudgetType = 'GLOBAL' | 'CATEGORY' | 'TASK_TYPE' | 'MODEL' | 'USER'

export type LimitRequestStatus = 'PENDING' | 'APPROVED' | 'DENIED'

// === Analytics Types ===

export interface LLMUsageSummary {
  total_requests: number
  total_tokens: number
  total_prompt_tokens: number
  total_completion_tokens: number
  total_cost_cents: number
  avg_duration_ms: number
  error_count: number
  error_rate: number
}

export interface LLMUsageByModel {
  model: string
  provider: LLMProvider
  request_count: number
  total_tokens: number
  cost_cents: number
  avg_tokens_per_request: number
}

export interface LLMUsageByTask {
  task_type: LLMTaskType
  request_count: number
  total_tokens: number
  cost_cents: number
  avg_duration_ms: number
}

export interface LLMUsageByCategory {
  category_id: string | null
  category_name: string | null
  request_count: number
  total_tokens: number
  cost_cents: number
}

export interface LLMUsageTrend {
  date: string
  request_count: number
  total_tokens: number
  cost_cents: number
  error_count: number
}

export interface LLMUsageTopConsumer {
  task_name: string
  task_type: LLMTaskType
  request_count: number
  total_tokens: number
  cost_cents: number
}

export interface LLMUsageByUser {
  user_id: string | null
  user_email: string | null
  user_name: string | null
  request_count: number
  total_tokens: number
  prompt_tokens: number
  completion_tokens: number
  cost_cents: number
  models_used: string[]
  has_credentials: boolean
}

export interface LLMUsageAnalytics {
  period_start: string
  period_end: string
  summary: LLMUsageSummary
  by_model: LLMUsageByModel[]
  by_task: LLMUsageByTask[]
  by_category: LLMUsageByCategory[]
  by_user: LLMUsageByUser[]
  daily_trend: LLMUsageTrend[]
  top_consumers: LLMUsageTopConsumer[]
}

export interface LLMCostProjection {
  current_month_cost_cents: number
  projected_month_cost_cents: number
  daily_avg_cost_cents: number
  budget_warning: boolean
  budget_limit_cents: number | null
}

// === Budget Types ===

export interface LLMBudgetConfig {
  id: string
  name: string
  budget_type: BudgetType
  reference_id: string | null
  reference_value: string | null
  monthly_limit_cents: number
  warning_threshold_percent: number
  critical_threshold_percent: number
  alert_emails: string[]
  is_active: boolean
  blocks_on_limit: boolean
  last_warning_sent_at: string | null
  last_critical_sent_at: string | null
  description: string | null
  created_at: string
  updated_at: string
}

export interface BudgetStatus {
  budget_id: string
  budget_name: string
  budget_type: BudgetType
  monthly_limit_cents: number
  current_usage_cents: number
  usage_percent: number
  warning_threshold_percent: number
  critical_threshold_percent: number
  is_warning: boolean
  is_critical: boolean
  is_blocked: boolean
  blocks_on_limit: boolean
  projected_month_end_cents: number
}

export interface BudgetStatusList {
  budgets: BudgetStatus[]
  any_warning: boolean
  any_critical: boolean
  any_blocked: boolean
}

export interface BudgetAlert {
  id: string
  budget_id: string
  alert_type: 'warning' | 'critical'
  threshold_percent: number
  current_usage_cents: number
  budget_limit_cents: number
  usage_percent: number
  email_sent: boolean
  created_at: string
}

// === Request Types ===

export interface LLMBudgetCreateRequest {
  name: string
  budget_type: BudgetType
  reference_id?: string
  reference_value?: string
  monthly_limit_cents: number
  warning_threshold_percent?: number
  critical_threshold_percent?: number
  alert_emails?: string[]
  description?: string
  is_active?: boolean
}

export interface LLMBudgetUpdateRequest {
  name?: string
  monthly_limit_cents?: number
  warning_threshold_percent?: number
  critical_threshold_percent?: number
  alert_emails?: string[]
  description?: string
  is_active?: boolean
}

export interface LLMUsageQueryParams {
  period?: '24h' | '7d' | '30d' | '90d'
  provider?: LLMProvider
  model?: string
  task_type?: LLMTaskType
  category_id?: string
}

// === Utility Types ===

export type PeriodOption = {
  label: string
  value: '24h' | '7d' | '30d' | '90d'
}

export const PERIOD_OPTIONS: PeriodOption[] = [
  { label: '24 Stunden', value: '24h' },
  { label: '7 Tage', value: '7d' },
  { label: '30 Tage', value: '30d' },
  { label: '90 Tage', value: '90d' },
]

export const PROVIDER_LABELS: Record<LLMProvider, string> = {
  AZURE_OPENAI: 'Azure OpenAI',
  OPENAI: 'OpenAI',
  ANTHROPIC: 'Anthropic Claude',
}

export const TASK_TYPE_LABELS: Record<LLMTaskType, string> = {
  SUMMARIZE: 'Zusammenfassen',
  EXTRACT: 'Extrahieren',
  CLASSIFY: 'Klassifizieren',
  EMBEDDING: 'Embeddings',
  VISION: 'Bildanalyse',
  CHAT: 'Chat',
  PLAN_MODE: 'Plan Mode',
  DISCOVERY: 'Discovery',
  ENTITY_ANALYSIS: 'Entity-Analyse',
  ATTACHMENT_ANALYSIS: 'Anhang-Analyse',
  RELEVANCE_CHECK: 'Relevanz-Check',
  CUSTOM: 'Benutzerdefiniert',
}

export const BUDGET_TYPE_LABELS: Record<BudgetType, string> = {
  GLOBAL: 'Global',
  CATEGORY: 'Pro Kategorie',
  TASK_TYPE: 'Pro Task-Typ',
  MODEL: 'Pro Modell',
  USER: 'Pro Benutzer',
}

// === User Budget Types ===

export interface UserBudgetStatus {
  budget_id: string
  monthly_limit_cents: number
  current_usage_cents: number
  usage_percent: number
  is_warning: boolean
  is_critical: boolean
  is_blocked: boolean
}

export interface LimitIncreaseRequest {
  id: string
  user_id: string
  budget_id: string
  requested_limit_cents: number
  current_limit_cents: number
  reason: string
  status: LimitRequestStatus
  reviewed_by: string | null
  reviewed_at: string | null
  admin_notes: string | null
  created_at: string
  user_email?: string | null
}

export interface LimitIncreaseRequestCreate {
  requested_limit_cents: number
  reason: string
}

export interface LimitRequestListResponse {
  requests: LimitIncreaseRequest[]
  total: number
  pending_count: number
}

export const LIMIT_REQUEST_STATUS_LABELS = {
  PENDING: 'Ausstehend',
  APPROVED: 'Genehmigt',
  DENIED: 'Abgelehnt',
} satisfies Record<LimitRequestStatus, string>

File: backend/app/schemas/llm_usage.py
"""Schemas for LLM usage analytics API."""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.llm_usage import LLMProvider, LLMTaskType

# === Response Models ===


class LLMUsageSummary(BaseModel):
    """Summary statistics for a time period."""

    total_requests: int = Field(description="Total number of API requests")
    total_tokens: int = Field(description="Total tokens consumed")
    total_prompt_tokens: int = Field(description="Total input tokens")
    total_completion_tokens: int = Field(description="Total output tokens")
    total_cost_cents: int = Field(description="Total estimated cost in USD cents")
    avg_duration_ms: float = Field(description="Average request duration in milliseconds")
    error_count: int = Field(description="Number of failed requests")
    error_rate: float = Field(description="Error rate as decimal (0-1)")


class LLMUsageByModel(BaseModel):
    """Usage breakdown by model."""

    model: str = Field(description="Model name/deployment")
    provider: LLMProvider = Field(description="LLM provider")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    cost_cents: int = Field(description="Estimated cost in USD cents")
    avg_tokens_per_request: float = Field(description="Average tokens per request")


class LLMUsageByTask(BaseModel):
    """Usage breakdown by task type."""

    task_type: LLMTaskType = Field(description="Type of task")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    cost_cents: int = Field(description="Estimated cost in USD cents")
    avg_duration_ms: float = Field(description="Average request duration")


class LLMUsageByCategory(BaseModel):
    """Usage breakdown by category."""

    category_id: UUID | None = Field(description="Category ID (null for uncategorized)")
    category_name: str | None = Field(description="Category name")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    cost_cents: int = Field(description="Estimated cost in USD cents")


class LLMUsageTrend(BaseModel):
    """Daily usage trend data point."""

    day: date = Field(description="Date of the data point")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    cost_cents: int = Field(description="Estimated cost in USD cents")
    error_count: int = Field(description="Number of errors")


class LLMUsageTopConsumer(BaseModel):
    """Top token consumers."""

    task_name: str = Field(description="Task name")
    task_type: LLMTaskType = Field(description="Task type")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    cost_cents: int = Field(description="Estimated cost in USD cents")


class LLMUsageByUser(BaseModel):
    """Usage breakdown by user (API credential owner)."""

    user_id: UUID | None = Field(description="User ID (null for system/legacy usage)")
    user_email: str | None = Field(description="User email")
    user_name: str | None = Field(description="User full name")
    request_count: int = Field(description="Number of requests")
    total_tokens: int = Field(description="Total tokens consumed")
    prompt_tokens: int = Field(description="Total input tokens")
    completion_tokens: int = Field(description="Total output tokens")
    cost_cents: int = Field(description="Estimated cost in USD cents")
    models_used: list[str] = Field(default_factory=list, description="List of models used")
    has_credentials: bool = Field(default=False, description="Whether user has API credentials configured")


class LLMUsageAnalyticsResponse(BaseModel):
    """Complete analytics response."""

    period_start: datetime = Field(description="Start of the analysis period")
    period_end: datetime = Field(description="End of the analysis period")
    summary: LLMUsageSummary = Field(description="Summary statistics")
    by_model: list[LLMUsageByModel] = Field(description="Breakdown by model")
    by_task: list[LLMUsageByTask] = Field(description="Breakdown by task type")
    by_category: list[LLMUsageByCategory] = Field(default_factory=list, description="Breakdown by category")
    by_user: list[LLMUsageByUser] = Field(default_factory=list, description="Breakdown by user")
    daily_trend: list[LLMUsageTrend] = Field(description="Daily usage trend")
    top_consumers: list[LLMUsageTopConsumer] = Field(description="Top token-consuming tasks")


class LLMCostProjection(BaseModel):
    """Cost projection based on current usage."""

    current_month_cost_cents: int = Field(description="Current month cost so far")
    projected_month_cost_cents: int = Field(description="Projected cost by end of month")
    daily_avg_cost_cents: int = Field(description="Average daily cost")
    budget_warning: bool = Field(default=False, description="Whether any budget warning is active")
    budget_limit_cents: int | None = Field(default=None, description="Global budget limit if set")


# === Request/Query Models ===


class LLMUsageQueryParams(BaseModel):
    """Query parameters for analytics endpoints."""

    period: str = Field(
        default="7d",
        description="Time period: 24h, 7d, 30d, 90d",
        pattern="^(24h|7d|30d|90d)$",
    )
    provider: LLMProvider | None = Field(default=None, description="Filter by provider")
    model: str | None = Field(default=None, description="Filter by model")
    task_type: LLMTaskType | None = Field(default=None, description="Filter by task type")
    category_id: UUID | None = Field(default=None, description="Filter by category")


class LLMUsageExportParams(BaseModel):
    """Parameters for exporting usage data."""

    period: str = Field(default="30d", description="Time period to export")
    format: str = Field(
        default="csv",
        description="Export format: csv, json",
        pattern="^(csv|json)$",
    )
    include_metadata: bool = Field(default=False, description="Include additional metadata fields")
